include/vast/ABI/MLIRTypeInfo.hpp

```diff
@@ -0,0 +1,252 @@
+
+        // TODO: Move into different abstraction, pointer size is dependent on underlying
+        //       cpu architecture not the IR.
+        static std::size_t pointer_size();
+
+        std::size_t size(mlir_type t);
+        std::size_t align(mlir_type t);
+
+        mlir_type iN(std::size_t s);
+        mlir_type fN(std::size_t s);
+
+        static mlir_type as_pointer(mlir_type t);
+
+        mlir_type type_at_offset(mlir_type t, std::size_t offset, const auto &accept);
+
+        auto field_containing_offset(mlir_type t, std::size_t offset)
+            -> std::optional< std::tuple< mlir_type, std::size_t > >;
+
+        gap::generator< mlir_type > mock_array_fields(hl::ArrayType array_type);
+
+        auto fields(mlir_type type);
+
+        using array_info_t = std::tuple< std::optional< std::size_t >, mlir_type >;
+        array_info_t array_info(mlir_type t);
+    };
+
+    bool vast_type_info::is_void(mlir_type t) {
+        return mlir::isa< mlir::NoneType >(t);
+    }
+
+    bool vast_type_info::is_scalar(mlir_type t) {
+        // TODO(abi): Also complex is an option in ABI.
+        return !is_aggregate(t);
+    }
+
+    bool vast_type_info::is_complex(mlir_type t) {
+        VAST_UNREACHABLE("Missing support for complex types!");
+    }
+
+    // TODO(abi): `SubElementTypeInterface` is not good enough, since for
+    //             example hl::PointerType implements it.
+    // TODO(abi): Figure how to better handle this than manual listing.
+    bool vast_type_info::is_aggregate(mlir_type t) {
+        return mlir::isa< hl::RecordType, hl::ElaboratedType, hl::ArrayType >(t);
+    }
+
+    bool vast_type_info::is_record(mlir_type t) {
+        if (t.isa< hl::RecordType >()) {
+            return true;
+        }
+        if (auto et = t.dyn_cast< hl::ElaboratedType >()) {
+            return is_record(et.getElementType());
+        }
+        return false;
+    }
+
+    bool vast_type_info::is_struct(mlir_type t) {
+        // TODO(abi): Are these equivalent?
+        return is_record(t);
+    }
+
+    bool vast_type_info::is_array(mlir_type t) {
+        return mlir::isa< hl::ArrayType >(hl::strip_elaborated(t));
+    }
+
+    // Must be invoked with `t` that is array-like.
+    auto vast_type_info::array_info(mlir_type t) -> array_info_t {
+        auto array_type = mlir::dyn_cast< hl::ArrayType >(t);
+        VAST_ASSERT(array_type);
+        return { array_type.getSize(), array_type.getElementType() };
+    }
+
+    bool vast_type_info::can_be_passed_in_regs(mlir_type t) {
+        // TODO(abi): Seems like in C nothing can prevent this.
+        return true;
+    }
+
+    bool vast_type_info::is_scalar_integer(mlir_type t) {
+        return t.isa< mlir::IntegerType >();
+    }
+
+    // TODO(abi): Implement.
+    bool vast_type_info::has_unaligned_field(mlir_type t) {
+        return false;
+    }
+
+    bool vast_type_info::is_scalar_float(mlir_type t) {
+        return t.isa< mlir::FloatType >();
```

Comment:
```suggestion
        return mlir::isa< mlir::FloatType >(t);
```

Reply:
what about `hl` float types?

Reply:
Already not present in this layer in the VAST conversion pipelines. Technically we probably should support it, but it is missing at more than just this place. I will open an issue - ideally we would want to have a separate class for just VAST-only types and one for upstream types.